windowing/egl: add Mali support (for Allwinner / Cubieboard)

MaxKellermann opened the pull request:
Add the class CEGLNativeTypeMali which uses the Mali binary driver
(https://github.com/linux-sunxi/sunxi-mali) to initialize an EGL
context.

This was tested successfully on a CubieTruck and a Cubieboard.

The screen resolution is currently hard-coded to the size of my TFT
monitor.  This part will need to be rewritten.

MaxKellermann committed:
windowing/egl: add Mali support

Add the class CEGLNativeTypeMali which uses the Mali binary driver
(https://github.com/linux-sunxi/sunxi-mali) to initialize an EGL
context.

This was tested successfully on a CubieTruck and a Cubieboard.

The screen resolution is currently hard-coded to the size of my TFT
monitor.  This part will need to be rewritten.

koying commented:
There doesn't seem to anything specific to Mali, here, just generic EGL.
It's good to have a "generic" egl implem as a fallback (i.e. should be tried last), but no need to put "mali" everywhere, then ;)

MaxKellermann commented:
Not true. "struct mali_native_window" is obviously a Mali specific data type that is initialized by my code in a Mali-specific way.

fritsch commented:
I am not sure I like the idea to support GPL violator hardware in GPL code ... basically I am quite sure I don't like that at all.

fritsch commented:
From the code: You basically hard code mali_native_window  which is kind of "magic" from an outside point of view. I think this code needs more steps.

As you said yourself the resolution handling is nowhere near ready and is just hardcoding.

fritsch commented:
mali_native_window <- there is no suspicious include at all, no idea where this is even coming from - someone patching EGL.h on those platforms?

koying commented:
@MaxKellermann Please reopen a PR when you have a complete implementation.
As it is, it's pretty much useless.

Thanks.

MaxKellermann commented:
Is there a point in spending time on completing the implementation? Would it be merged? If yes, what is missing from your point of view?

koying commented:
1) We don't see how it works. Just putting some hardcoded values to a struct we know nothing about doesn't tell us how it's supposed to work
2) Obviously using non-standard hardcoded window size is not valid

On what was this tested?

koying commented:
Oh, and obviously TARGET_MALI must be enabled somewhere (configure.ac) or this is just dead code

koying commented:
Note that is the OS is Android, your code is just never used, as CEGLNativeTypeAndroid will always be picked up first.

MaxKellermann commented:
"We don't see how it works" - Was that supposed to be the question "How does this work?" It works by using the Mali EGL user-space API in a documented way.

"Obviously using non-standard hardcoded window size is not valid" - true, my comment says exactly that. I wanted to be sure my work is going somewhere before I throw more of my time at it. I'm not yet convinced that the Kodi project is interested.

"On what was this tested?" - You can read that in my commit message: "This was tested successfully on a CubieTruck and a Cubieboard."

"Oh, and obviously TARGET_MALI must be enabled somewhere (configure.ac) or this is just dead code (hardcoding it is obviously not an option)" Not exactly true. From my patch: "+#define TARGET_MALI // TODO: define somewhere else?"

"Note that if the OS is Android, your code is just never used, as CEGLNativeTypeAndroid will always be picked up first." -  On Android, this patch is useless because it would circumvent Android's SurfaceFlinger. So this is about "plain" Linux (X11 or frame buffer). I didn't think it was necessary to point that out.

koying commented:
OK, please point to the documentation you mention. Not sure why you seem to consider this is common knowledge. 

Is it egl over framebuffer or egl over x11? If the latter, what's wrong with the existing implementation? If the former, we definitely need the doc to understand the magic of not even mentioning a fb device. 

I suppose it only builds with the Mali egl.h? How do you plan to include it in our cross-compile environment (tools/depend/target)? 
How do you build it yourself? On device? 

Generally speaking, if your PR is wip, please say so. Not quite sure what you expected by submitting an early work, with little to no explanations, and let it rot for 3 months without pinging anyone.

MaxKellermann commented:
"OK, please point to the documentation you mention" - https://github.com/linux-sunxi/sunxi-mali/blob/master/include/EGL/eglplatform_fb.h and https://github.com/linux-sunxi/sunxi-mali/blob/master/test/test.c#L42

"Is it egl over framebuffer or egl over x11?" - Only framebuffer. Using X11 adds an unnecessary layer of overhead. (My previous comment here about this mentioning X11 was wrong.)

"I suppose it only builds with the Mali egl.h?" - Yes, just like **all** uses of EGL require a implementation-specific EGL headers, with implementation-specific "native" data structures.

"How do you plan to include it in our cross-compile environment (tools/depend/target)?" - I don't know how your cross-compile environment works. Where's documentation about it? What is usually done in your project to depend on non-free system libraries?

"How do you build it yourself? On device?" - Yes.

"Not quite sure what you expected by submitting an early work, with little to no explanations, and let it rot for 3 months without pinging anyone." - Let it rot? Huh? Everybody else let it rot. I don't beg to merge this, and I don't think my code is too important. I just submitted it because somebody may find it useful, but I'm happy with running my own Kodi fork.

MaxKellermann commented:
Rebased on current master.

koying commented:
Thanks, much more clear, now.

As you might have noticed, Allwinner is a difficult topic at Kodi, but I'd still suggest that you complete your implementation, even if it probably won't be merged, for the benefit of other sunxi users.
If the PR title is clear, it will be easy for them to find and cherrypick it.

OTOH, I think I remember the sunxi Mali egl is used on other SoC's with Mali 4xx (Odroid Exynos, IIRC). If so, it might actually be valuable to merge.

Suggestions:
- Use "Sunxi" or "SunxiMali" in class names, macro, ... to make it clear it's not generic Mali
- Use "USE_" rather than "TARGET_" (TARGET means full platform, internally, while USE is for a specific feature)
- Implement fb size detection, to have a working implem
- Let's assume that this will only be built on device (or buildroot). So, in configure.ac, detect that the system <EGL/eglplatform.h> is the sunxi one, somehow, and enable "USE_SUNXIEGL" accordingly.
- Be sure that none of your code is built for non-sunxi platforms (see recent change to EGLWrapper in trunk) for easier acceptation.

MaxKellermann commented:
Pushed amended commit: https://github.com/MaxKellermann/xbmc/commit/7b37e76003490e854195db0db4c6c3ebde30b002

Changes as requested by koying:
- added "Sunxi" prefix
- "USE_" instead of "TARGET_"
- detect the Sunxi Mali EGL header
- added #ifdef in EGLWrapper

Frame buffer size detection not yet implemented, and that will be the last missing piece (AFAIK).
